### DISCUSSION TITLE
Handling pseudo terminal newlines

### DISCUSSION THREAD

**Discussion by fswarbrick** [zopencommunity/meta#812] [Q&A]
Edit:  Of course once I post I find the answer.  Specifically, I found the following in https://github.com/ZOSOpenTools/tmuxport/blob/main/patches/spawn.c.patch#L18:
```
+#endif
+#ifdef __MVS__
+  now.c_oflag |= (OPOST|ONLCR);
```
I could swear I tried that already, but I'm thinking I applied it to the wrong "side" of the pseudo-terminal.

Anyway, it seems to be resolved now.  I leave the original message below for historical purposes.

So z/OS Unix has an interesting...characteristic.  On other POSIX/*nix platforms when an application is writing to a terminal, any time a linefeed/newline ('\n') is to be written to the terminal, a carriage return is added in front of it.  This causes the terminal to both start on a new line, **and** at the beginning of that line.  Where this logic is performed I do not know.  On z/OS, however, this does not happen (when using the ASCII APIs).  I am thinking it's because of conversion between EBCDIC and ASCII, where z/OS uses an EBCDIC NL character, which is different from ASCII LF, but appears to be converted from NL to LF.

While this is correct behavior when writing to a file, when writing to a terminal it causes the cursor on the new line to not be moved to the first row, but rather starts printing just below where the previous line ended.

It looks like z/OS accounts for this when writing to a terminal.  That is, if you write a line ending with only a newline character to a terminal it appears to add a carriage return before the LF, and thus the terminal behaves appropriately.  In any (?) other case it simply writes the LF.

The issue is with a terminal emulator program, written in Go (though I'm not sure that is significant).  The application in question is the Micro TUI editor.  It has a feature to open up a pseudo terminal in a program window.  After adding z/OS support to Micro and it's required packages, it all works great except that the "NL w/o CR" behavior occurs.  This is because, I believe, Micro uses ANSI/VT100 escape codes when writing to the terminal, rather than writing the raw newline.

There is a code within the Terminal package that Micro uses that turns on "ModeCRLF".  When this flag is true, the Terminal packages tells the screen manager to, in addition to going down one row, also go to the beginning of that row.

I have a solution that works, but I am not happy with it.  I am wondering how it is that other "terminal emulation" programs (screen, tmux and Vim) seem to be able to be ported without any special logic to handle this, but Micro seems to require special logic.

Basically, I added code to Micro to place the terminal emulator program in "new line mode" (in contrast to the default of "line feed mode" by writing a VT100 "LNM" (Set new line mode) escape code ("\033[20h") (see https://espterm.github.io/docs/VT100%20escape%20codes.html) before it starts it's I/O loop with the master side of the pseudo-terminal.  I see no evidence that these other programs do this, so I'm wondering how it is that they work without it.

Any ideas?  Below are some code snippets from the Terminal package and from Micro.  I added the line sending the LNM code in the Micro function "(t *Terminal) Start()".

**https://github.com/zyedidia/terminal**
```
parse.go
func (t *State) handleControlCodes(c rune) bool {
	[...]
	switch c {
	[...]
	// LF, VT, LF
	case '\f', '\v', '\n':
		// go to first col if mode is set
		t.newline(t.mode&ModeCRLF != 0)
	[...]
	}
	[...]
}

state.go
func (t *State) setMode(priv bool, set bool, args []int) {
	if priv {
			[...]
	} else {
		for _, a := range args {
			switch a {
			[...]
			case 20: // LNM - linefeed/newline
				t.modMode(set, ModeCRLF)
			[...]
			default:
				t.logf("unknown set/reset mode %d\n", a)
			}
		}
	}
}

func (t *State) newline(firstCol bool) {
	y := t.cur.y
	if y == t.bottom {
		cur := t.cur
		t.cur = t.defaultCursor()
		t.ScrollUp(t.top, 1)
		t.cur = cur
	} else {
		y++
	}
	if firstCol {
		t.moveTo(0, y)
	} else {
		t.moveTo(t.cur.x, y)
	}
}
```
**https://github.com/zyedidia/micro**
```
// Start begins a new command in this terminal with a given view
func (t *Terminal) Start(execCmd []string, getOutput bool, wait bool, callback func(out string, userargs []interface{}), userargs []interface{}) error {
        if len(execCmd) <= 0 {
                return nil
        }

        cmd := exec.Command(execCmd[0], execCmd[1:]...)
        t.output = nil
        if getOutput {
                t.output = bytes.NewBuffer([]byte{})
                cmd.Stdout = t.output
        }
        Term, _, err := terminal.Start(&t.State, cmd)
        if err != nil {
                return err
        }
        t.Term = Term
        t.getOutput = getOutput
        t.Status = TTRunning
        t.title = execCmd[0] + ":" + strconv.Itoa(cmd.Process.Pid)
        t.wait = wait
        t.callback = func(out string) {
                callback(out, userargs)
        }

        t.Term.Write([]byte("\033[20h")) // set CRLF mode (for z/OS only)

        go func() {
                for {
                        err := Term.Parse()
                        if err != nil {
                                Term.Write([]byte("Press enter to close"))
                                screen.Redraw()
                                break
                        }
                        screen.Redraw()
                }
                t.Stop()
        }()

        return nil
}
```

**Comment by IgorTodorovskiIBM** [comment]
That one-line patch took me days to figure out! Glad you got it working
